File: LUISBot/node/contentDialog.js
const { LuisRecognizer } = require("botbuilder-ai");
const { CardFactory, MessageFactory } = require("botbuilder");

const {
  ChoiceFactory,
  ChoicePrompt,
  ComponentDialog,
  ConfirmPrompt,
  DialogSet,
  DialogTurnStatus,
  TextPrompt,
  WaterfallDialog,
} = require("botbuilder-dialogs");

const CHOICE_PROMPT = "CHOICE_PROMPT";
const CONFIRM_PROMPT = "CONFIRM_PROMPT";
const LANG_PROMPT = "LANG_PROMPT";
const WATERFALL_DIALOG = "WATERFALL_DIALOG";

const MS_Root = "https://docs.microsoft.com/";
const MS_SamplesSearchURL = "samples/browse/?terms=";
const MS_DocsSearchURL = "search/?terms=";

class ContentDialog extends ComponentDialog {
  constructor(luis) {
    super("contentDialog");
    this.luis = luis;

    this.addDialog(new ChoicePrompt(CHOICE_PROMPT));
    this.addDialog(new TextPrompt(LANG_PROMPT));
    this.addDialog(new ConfirmPrompt(CONFIRM_PROMPT));

    this.addDialog(
      new WaterfallDialog(WATERFALL_DIALOG, [
        this.selectResourceStep.bind(this),
        this.selectTechStep.bind(this),
        this.confirmChoices.bind(this),
        this.sendResults.bind(this),
      ])
    );

    this.initialDialogId = WATERFALL_DIALOG;
  }

  async selectResourceStep(step) {
    if (!this.luis.isConfigured) {
      const messageText =
        "NOTE: LUIS is not configured. To enable all capabilities, add `LuisAppId`, `LuisAPIKey` and `LuisAPIHostName` to the .env file.";
      step.context.sendActivity(messageText);
      return await step.endDialog();
    } else {
      return await step.prompt(CHOICE_PROMPT, {
        prompt: "Which learning resources did you want?",
        choices: ChoiceFactory.toChoices(["Docs", "Samples", "Both"]),
      });
    }
  }

  async selectTechStep(step) {
    step.values.resources = step.result.value;
    return await step.prompt(LANG_PROMPT, "What language did you want?");
  }

  async confirmChoices(step) {
    step.values.lang = step.result;
    const luisResult = await this.luis.executeLuisQuery(step.context);
    const message = `You want to see ${step.values.resources} in ${step.values.lang} Is this correct?`;

    if (LuisRecognizer.topIntent(luisResult) != "Learning") {
      step.context.sendActivity(
        `Your query for ${step.values.lang} was not recognized by this bot, but will still be added to the search results`
      );
    }

    return await step.prompt(CONFIRM_PROMPT, message);
  }

  async sendResults(step) {
    let url = MS_Root;
    let resource = step.values.resources;
    let language = step.values.lang;

    let sample = {
      type: "Action.OpenUrl",
      title: "MS Samples on " + language,
      url: `${url + MS_SamplesSearchURL + language}`,
    };

    let docs = {
      type: "Action.OpenUrl",
      title: "MS Docs on " + language,
      url: `${url + MS_DocsSearchURL + language}`,
    };

    let cardSchema = {
      $schema: "http://adaptivecards.io/schemas/adaptive-card.json",
      type: "AdaptiveCard",
      version: "1.0",
      body: [
        {
          type: "TextBlock",
          text: "Here's your links!",
        },
      ],
      actions: [],
    };

    if (resource == "Docs") cardSchema.actions.push(docs);
    else if (resource == "Samples") cardSchema.actions.push(sample);
    else {
      cardSchema.actions.push(docs);
      cardSchema.actions.push(sample);
    }

    const card = CardFactory.adaptiveCard(cardSchema);
    const message = MessageFactory.attachment(card);
    await step.context.sendActivity(message);

    return await step.endDialog();
  }

  async run(turnContext, accessor) {
    const dialogSet = new DialogSet(accessor);
    dialogSet.add(this);
    const dialogContext = await dialogSet.createContext(turnContext);
    const results = await dialogContext.continueDialog();
    if (results.status === DialogTurnStatus.empty) {
      await dialogContext.beginDialog(this.id);
    }
  }
}

module.exports.ContentDialog = ContentDialog;

File: LUISBot/dotnet_core/ContentDialog.cs
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Choices;
using Microsoft.Bot.Schema;
using Microsoft.Extensions.Logging;

namespace LUISBotSample
{
    public class ContentDialog : ComponentDialog
    {
        private readonly ContentRecognizer _luisRecognizer;
        protected readonly ILogger Logger;

       private readonly string  MS_Root = "https://docs.microsoft.com/";
       private readonly string  MS_SamplesSearchURL = "samples/browse/?terms=";
       private readonly string  MS_DocsSearchURL = "search/?terms=";

        public string DocsUrl => MS_Root + MS_DocsSearchURL;
        public string SamplesUrl => MS_Root + MS_SamplesSearchURL;

        public ContentDialog(ContentRecognizer luisRecognizer, ILogger<ContentDialog> logger)
                  : base(nameof(ContentDialog))
        {
            _luisRecognizer = luisRecognizer;
            Logger = logger;


            AddDialog(new TextPrompt(nameof(TextPrompt)));
            AddDialog(new ChoicePrompt(nameof(ChoicePrompt)));
            AddDialog(new ConfirmPrompt(nameof(ConfirmPrompt)));

            AddDialog(new WaterfallDialog(nameof(WaterfallDialog), new WaterfallStep[]
            {
                SelectResourceStepAsync,
                SelectTechStepAsync,
                ConfirmChoicesStepAsync,
                SendResults,
            }));

            // The initial child Dialog to run.
            InitialDialogId = nameof(WaterfallDialog);
        }
        private async Task<DialogTurnResult> SelectResourceStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            if (!_luisRecognizer.IsConfigured)
            {
                await stepContext.Context.SendActivityAsync(
                    MessageFactory.Text("LUIS is not configured. To enable all capabilities, add 'LuisAppId', 'LuisAPIKey' and 'LuisAPIHostName' to the appsettings.json file."), cancellationToken);

                return await stepContext.NextAsync(null, cancellationToken);
            }

            return await stepContext.PromptAsync(nameof(ChoicePrompt),
            new PromptOptions
            {
                Prompt = MessageFactory.Text("Welcome to the Docs Bot! Which learning resources did you want?"),
                Choices = ChoiceFactory.ToChoices(new List<string> { "Docs", "Samples", "Both" }),
            }, cancellationToken);
        }

        private async Task<DialogTurnResult> SelectTechStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            stepContext.Values["resource"] = ((FoundChoice)stepContext.Result).Value;
            return await stepContext.PromptAsync(nameof(TextPrompt), new PromptOptions { Prompt = MessageFactory.Text("What language did you want?") }, cancellationToken);
        }

        private async Task<DialogTurnResult> ConfirmChoicesStepAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            stepContext.Values["lang"] = (string)stepContext.Result;
            var resource = (string)stepContext.Values["resource"];
            var lang = (string)stepContext.Values["lang"];

            var luisResult = await _luisRecognizer.RecognizeAsync<ContentModel>(stepContext.Context, cancellationToken);
            if (luisResult.TopIntent().intent != ContentModel.Intent.Learning)
            {
                MessageFactory.Text($"Your query for {lang} was not recognized by this bot, but will still be added to the search results");
            }
            return await stepContext.PromptAsync(nameof(ConfirmPrompt), new PromptOptions { Prompt = MessageFactory.Text($"You want to see {resource} in {lang} Is this correct?") }, cancellationToken);
        }

        private async Task<DialogTurnResult> SendResults(WaterfallStepContext stepContext, CancellationToken cancellationToken)
        {
            var attachments = new List<Attachment>();
            var reply = MessageFactory.Attachment(attachments);

            var resource = (string)stepContext.Values["resource"];
            var lang = (string)stepContext.Values["lang"];

            var cardActions = new List<CardAction>();

            if (resource == "Docs" || resource == "Both")
                cardActions.Add(new CardAction(ActionTypes.OpenUrl, $"MS Docs on {lang}", value: DocsUrl + lang));
            if (resource == "Samples" || resource == "Both")
                cardActions.Add(new CardAction(ActionTypes.OpenUrl, $"MS Samples on {lang}", value: SamplesUrl + lang));

            var adaptiveCardAttachment = new HeroCard()
            {
                Text = "Here's your links!",
                Buttons = cardActions
            };

            reply.Attachments.Add(adaptiveCardAttachment.ToAttachment());
            await stepContext.Context.SendActivityAsync(reply, cancellationToken);

            return await stepContext.EndDialogAsync();
        }
    }
}